Initial documentation for System.IO.Compression.BrotliStream, amending GZipStream/DeflateStream

carlossanlop opened the pull request:
## Summary

Add initial documentation to System.IO.Compression.BrotliStream (it currently has none).
These updates were based on similar classes: GZipStream and DeflateStream, which are mostly documented.
Note: This change does not contain any remarks. I am reviewing the existing ones in GZipStream/DeflateStream to see if they also apply to BrotliStream. If they do, I will add them in a different iteration.

@mairaw @karelz PTAL

dnfclas commented:
[![CLA assistant check](https://cla.dotnetfoundation.org/pull/badge/not_signed)](https://cla.dotnetfoundation.org/dotnet/dotnet-api-docs?pullRequest=2001) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/dotnet/dotnet-api-docs?pullRequest=2001) before we can accept your contribution.<br/><br/>:x: carlossanlop [sign now](https://cla.dotnetfoundation.org/dotnet/dotnet-api-docs?pullRequest=2001)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/dotnet/dotnet-api-docs?pullRequest=2001) it.</sub>

mairaw commented on xml/System.IO.Compression/BrotliStream.xml:
```diff
@@ -34,9 +34,9 @@
-        <summary>To be added.</summary>
+        <param name="stream">The stream to compress.</param>
+        <param name="compressionLevel">One of the enumeration values that indicates whether to emphasize speed or compression efficiency when compressing the stream.</param>
+        <summary>Initializes a new instance of the <see. cref="T:System.IO.Compression.Brotli.BrotliStream" /> class by using the specified stream and compression level.</summary>
```
```suggestion
        <summary>Initializes a new instance of the <see cref="T:System.IO.Compression.Brotli.BrotliStream" /> class by using the specified stream and compression level.</summary>
```

mairaw commented on xml/System.IO.Compression/BrotliStream.xml:
```diff
@@ -56,9 +56,9 @@
-        <summary>To be added.</summary>
+        <param name="stream">The stream to compress.</param>
+        <param name="mode">One of the enumeration values that indicates whether to compress or decompress the stream.</param>
+        <summary>Initializes a new instance of the <see. cref="T:System.IO.Compression.Brotli.BrotliStream" /> class by using the specified stream and compression mode.</summary>
```
```suggestion
        <summary>Initializes a new instance of the <see cref="T:System.IO.Compression.Brotli.BrotliStream" /> class by using the specified stream and compression mode.</summary>
```

mairaw commented on xml/System.IO.Compression/BrotliStream.xml:
```diff
@@ -103,10 +103,10 @@
+        <param name="mode">One of the enumeration values that indicates whether to compress or decompress the stream.</param>
         <param name="leaveOpen">To be added.</param>
-        <summary>To be added.</summary>
+        <summary>Initializes a new instance of the <see. cref="T:System.IO.Compression.Brotli.BrotliStream" /> class by using the specified stream and compression mode, and optionally leaves the stream open.</summary>
```
```suggestion
        <summary>Initializes a new instance of the <see cref="T:System.IO.Compression.Brotli.BrotliStream" /> class by using the specified stream and compression mode, and optionally leaves the stream open.</summary>
```

carlossanlop replied:
Thank you for catching those extra dots, I totally missed them.

mairaw commented on xml/System.IO.Compression/BrotliStream.xml:
```diff
@@ -126,9 +126,10 @@
+        <summary>Gets a reference to the underlying stream.</summary>
+        <value>A stream object that represents the underlying stream.</value>
         <remarks>To be added.</remarks>
+        <exception cref="T:System.ObjectDisposedException">The underlying stream is closed.</exception> <!-- Question for reviewers: The actual exception message is 'Cannot access a closed file.'. Should I use that string instead? The string 'The underlying stream is closed' is used in GZipStream and DeflateStream and explains 'when' the exception is thrown, but is not the actual exception messag ...
```
You don't have to use the exact same error message. The docs should explain when you get that exception.

carlossanlop replied:
Thank you. I removed the comment with the question.

mairaw commented on xml/System.IO.Compression/BrotliStream.xml:
```diff
@@ -146,21 +147,27 @@
         <ReturnType>System.IAsyncResult</ReturnType>
       </ReturnValue>
       <Parameters>
+        <!-- Question for reviewers: GZipStream and DeflateStream have additional attributes in these Parameters: Index, FrameworkAlternate. Where those autogenerated or do I have to investigate them and add them manually myself? If it's the latter, where did the FrameworkAlternate info come from? I suspect it has something to do with the extra params 'cback' and 'state' which this function does  ...
```
I'll take a look but you shouldn't mess with those. They come from the mdoc reflection tool.

carlossanlop replied:
Gotcha. Thanks!

carlossanlop replied:
Question comment removed.

mairaw commented on xml/System.IO.Compression/BrotliStream.xml:
```diff
@@ -414,9 +430,10 @@
-        <remarks>To be added.</remarks>
+        <summary>This property is not supported and always throws a <see cref="T:System.NotSupportedException" />.</summary>
+        <value>A long value.</value>
+        <remarks>To be added.</value>
```
This is what's causing the build error.

```suggestion
        <remarks>To be added.</remarks>
```

carlossanlop replied:
Thanks!

mairaw commented on xml/System.IO.Compression/GZipStream.xml:
```diff
@@ -392,7 +392,7 @@
         <param name="asyncState">A user-provided object that distinguishes this particular asynchronous read request from other requests.</param>
         <param name="state">A user-provided object that distinguishes this particular asynchronous read request from other requests.</param>
-        <summary>Begins an asynchronous read operation. (Consider using the <see cref="M:System.IO.Stream.ReadAsync(System.Byte[],System.Int32,System.Int32)" /> method instead.)</summary>
+        <summary>Begins an asynchronous read operation. (Consider using the <see cref="M:System.IO.Compression.GZipStream.ReadAsync(System.Byte[],System.Int32,System.Int32)" /> method instead.)</summary>
```
ok, you got a couple of build warnings. This change is actually bad since the method on GZipStream is inherited from Stream, so you don't have API in the GZipStream class itself

carlossanlop replied:
Oh you're right, I can see that now, thanks for letting me know.

There are only two options in GZipStream:

```
ReadAsync(byte[] array, int offset, int count, CancellationToken cancellationToken);
ReadAsync(Memory<byte> buffer, CancellationToken cancellationToken = default(CancellationToken));
```

But there isn't one with (byte[], int, int).

I'll amend the code.

mairaw commented on xml/System.IO.Compression/BrotliStream.xml:
```diff
@@ -79,10 +79,10 @@
+        <param name="stream">The stream to compress.</param>
+        <param name="compressionLevel">One of the enumeration values that indicates whether to emphasize speed or compression efficiency when compressing the stream.</param>
+        <param name="leaveOpen"><see langword="true" /> to leave the stream open after disposing the <see cref="T:System.IO.Compression.Brotli.BrotliStream" /> object; otherwise, <see langword="false" />.</param>
+        <summary>Initializes a new instance of the <see. cref="T:System.IO.Compression.Brotli.BrotliStream" /> class by using the specified stream and compression level, and optionally leaves the stream open.</summary>
```
```suggestion
        <summary>Initializes a new instance of the <see cref="T:System.IO.Compression.Brotli.BrotliStream" /> class by using the specified stream and compression level, and optionally leaves the stream open.</summary>
```

mairaw commented:
You can now see the content on the review site here:
https://review.docs.microsoft.com/en-us/dotnet/api/system.io.compression.brotlistream?view=netcore-2.2&branch=pr-en-us-2001

And on the build report, you'll see the new warnings you got towards the bottom, on a section called **Validate Files**
https://opbuildstorageprod.blob.core.windows.net/report/2019%5C3%5C7%5C500c749d-de78-3d4a-00e4-269d599bf9ec%5CPullRequest%5C201903071857430570-2001%5Cworkflow_report.html?sv=2015-02-21&sr=b&sig=rGdSfkl%2FeOcpFAtdS2itTLz4tOp6oD2Gk1jvGi0ncSM%3D&st=2019-03-07T20%3A23%3A58Z&se=2019-04-07T20%3A28%3A58Z&sp=r

carlossanlop commented:
@mairaw Thank you for catching the ReadAsync issue. I corrected the same comment in all the files I modified, and made sure the suggestion to use ReadAsync/WriteAsync functions point to the System.IO.Stream implementations.

mairaw commented on xml/System.IO.Compression/BrotliStream.xml:
```diff
@@ -56,9 +56,9 @@
-        <summary>To be added.</summary>
+        <param name="stream">The stream to compress.</param>
+        <param name="mode">One of the enumeration values that indicates whether to compress or decompress the stream.</param>
+        <summary>Initializes a new instance of the <see cref="T:System.IO.Compression.Brotli.BrotliStream" /> class by using the specified stream and compression mode.</summary>
```
```suggestion
        <summary>Initializes a new instance of the <see cref="T:System.IO.Compression.BrotliStream" /> class by using the specified stream and compression mode.</summary>
```

mairaw commented on xml/System.IO.Compression/BrotliStream.xml:
```diff
@@ -34,9 +34,9 @@
-        <summary>To be added.</summary>
+        <param name="stream">The stream to compress.</param>
+        <param name="compressionLevel">One of the enumeration values that indicates whether to emphasize speed or compression efficiency when compressing the stream.</param>
+        <summary>Initializes a new instance of the <see cref="T:System.IO.Compression.Brotli.BrotliStream" /> class by using the specified stream and compression level.</summary>
```
```suggestion
        <summary>Initializes a new instance of the <see cref="T:System.IO.Compression.BrotliStream" /> class by using the specified stream and compression level.</summary>
```

mairaw commented on xml/System.IO.Compression/BrotliStream.xml:
```diff
@@ -79,10 +79,10 @@
-        <summary>To be added.</summary>
+        <param name="stream">The stream to compress.</param>
+        <param name="compressionLevel">One of the enumeration values that indicates whether to emphasize speed or compression efficiency when compressing the stream.</param>
+        <param name="leaveOpen"><see langword="true" /> to leave the stream open after disposing the <see cref="T:System.IO.Compression.Brotli.BrotliStream" /> object; otherwise, <see langword="false" />.</param>
```
```suggestion
        <param name="leaveOpen"><see langword="true" /> to leave the stream open after disposing the <see cref="T:System.IO.Compression.BrotliStream" /> object; otherwise, <see langword="false" />.</param>
```

mairaw commented on xml/System.IO.Compression/BrotliStream.xml:
```diff
@@ -79,10 +79,10 @@
+        <param name="stream">The stream to compress.</param>
+        <param name="compressionLevel">One of the enumeration values that indicates whether to emphasize speed or compression efficiency when compressing the stream.</param>
+        <param name="leaveOpen"><see langword="true" /> to leave the stream open after disposing the <see cref="T:System.IO.Compression.Brotli.BrotliStream" /> object; otherwise, <see langword="false" />.</param>
+        <summary>Initializes a new instance of the <see cref="T:System.IO.Compression.Brotli.BrotliStream" /> class by using the specified stream and compression level, and optionally leaves the stream open.</summary>
```
```suggestion
        <summary>Initializes a new instance of the <see cref="T:System.IO.Compression.BrotliStream" /> class by using the specified stream and compression level, and optionally leaves the stream open.</summary>
```

mairaw commented on xml/System.IO.Compression/BrotliStream.xml:
```diff
@@ -103,10 +103,10 @@
+        <param name="mode">One of the enumeration values that indicates whether to compress or decompress the stream.</param>
         <param name="leaveOpen">To be added.</param>
-        <summary>To be added.</summary>
+        <summary>Initializes a new instance of the <see cref="T:System.IO.Compression.Brotli.BrotliStream" /> class by using the specified stream and compression mode, and optionally leaves the stream open.</summary>
```
```suggestion
        <summary>Initializes a new instance of the <see cref="T:System.IO.Compression.BrotliStream" /> class by using the specified stream and compression mode, and optionally leaves the stream open.</summary>
```

carlossanlop replied:
Oops, thanks for catching that.

mairaw commented on xml/System.IO.Compression/BrotliStream.xml:
```diff
@@ -486,12 +503,18 @@
+        <exception cref="T:System.ObjectDisposedException">The underlying stream is null or closed.</exception>
       </Docs>
     </Member>
     <Member MemberName="ReadAsync">
```
There are still a few APIs missing docs. Do you plan to cover those on the next update? e.g. Line 538

carlossanlop replied:
Correct. I decided to start working with the APIs that had already been described in Stream, GZipStream and DeflateStream.
I will cover the rest of the BrotliStream, GZipStream and DeflateStream APIs in a future pull request. I'm currently investigating how they work.

carlossanlop commented:
Thank you!
